DRA: resource quotas

pohly opened the pull request:
#### What type of PR is this?

/kind feature
/kind api-change

#### What this PR does / why we need it:

Dynamic resource allocation is similar to storage in the sense that users
create ResourceClaim objects to request resources, same as with persistent
volume claims. The actual resource usage is only known when allocating claims,
but some limits can already be enforced at admission time:
    
- `count/resourceclaims.resource.k8s.io` limits the number of ResourceClaim objects in
  a namespace; this is a generic feature that is already supported also without
  this commit.

- "resourceclaims" is *not* an alias - use "count/resourceclaims.resource.k8s.io"
  instead.

- `<device-class-name>.deviceclass.resource.k8s.io/devices` limits the number of
  ResourceClaim objects in a namespace such that the number of devices
  requested through those objects with that class does not exceed the limit.
    
A single request may cause the allocation of multiple devices. For exact
counts, the quota limit is based on the sum of those exact counts. For requests
asking for "all" matching devices, the maximum number of allocated devices per
claim is used as a worst-case upper bound.
    
Requests asking for "admin access" contribute to the quota.

#### Which issue(s) this PR fixes:
Fixes https://github.com/kubernetes/kubernetes/issues/114916

#### Special notes for your reviewer:

This is derived from the persistent volume claim handler. To make it more obvious what was changed, the first commit is a 1:1 copy of that handler's implementation and unit test.

#### Does this PR introduce a user-facing change?

```release-note
DRA: the number of ResourceClaim objects can be limited per namespace and by the number of devices requested through a specific class via the v1.ResourceQuota mechanism.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/issues/3063
```

pohly committed:
resource quota: clone PVC quota evaluator for DRA

pohly committed:
DRA quota: add ResourceClaim v1.ResourceQuota limits

Dynamic resource allocation is similar to storage in the sense that users
create ResourceClaim objects to request resources, same as with persistent
volume claims. The actual resource usage is only known when allocating claims,
but some limits can already be enforced at admission time:

- "count/resourceclaims.resource.k8s.io" limits the number of ResourceClaim objects in
  a namespace; this is a generic feature that is already supported also without
  this commit.

- "resourceclaims" is *not* an alias - use "count/resourceclaims.resource.k8s.io"
  instead.

- <device-class-name>.deviceclass.resource.k8s.io/devices limits the number of
  ResourceClaim objects in a namespace such that the number of devices
  requested through those objects with that class does not exceed the limit.

A single request may cause the allocation of multiple devices. For exact
counts, the quota limit is based on the sum of those exact counts. For requests
asking for "all" matching devices, the maximum number of allocated devices per
claim is used as a worst-case upper bound.

Requests asking for "admin access" contribute to the quota.

DRA quota: remove admin mode exception

alexzielenski commented:
/triage accepted

pohly commented:
/retest

liggitt commented on pkg/apis/core/helper/helpers.go:
```diff
@@ -232,7 +234,8 @@ var standardQuotaResources = sets.NewString(
 // IsStandardQuotaResourceName returns true if the resource is known to
 // the quota tracking system
 func IsStandardQuotaResourceName(str string) bool {
```
The functions in this file are used in API validation and are *really* hard to reason about changes to, especially across skewed server versions.

A feature gate is not sufficient, or the first minor version the feature gate is enabled in will allow in data the previous version of the server considers invalid.

I think the functions we're changing here have to be parameterized, and informed by the feature gate OR call sites validating updates to existing objects already containing this additional resource type.

liggitt replied:
(it would not surprise me if this was done in an unsafe way when persistentvolumeclaim quota was added... we thought through the implications of skewed validation differences a lot more in the ~1.14 timeframe)

pohly replied:
I think I understand what you mean.

But suppose this goes onto 1.29, someone enables the feature gate, creates an object with a new name, and then downgrades. In 1.28 the name was not know and validation fails.

Does that mean we have to add some code now which accepts the names regardless of the feature gate *for existing objects* but not for *new objects*, then once all supported releases have that code, allow setting them in new objects? How many releases do we have to wait?

liggitt replied:
> But suppose this goes onto 1.29, someone enables the feature gate, creates an object with a new name, and then downgrades. In 1.28 the name was not know and validation fails.

If the gate is alpha and they opt-in and persist data which a 1.28 server considers invalid, that's on them. In the docs for this, state all servers must be 1.29+ before trying out the alpha feature. We would not promote the gate to beta or enable it by default until the oldest skewed server would recognize / tolerate the data.

> Does that mean we have to add some code now which accepts the names regardless of the feature gate for existing objects but not for new objects,

yes, that's what I was trying to describe in my "informed by the feature gate OR call sites validating updates to existing objects already containing this additional resource type" comment, sorry that wasn't clear.

> then once all supported releases have that code, allow setting them in new objects? How many releases do we have to wait?

we only have to wait a single release, so the n-1 server has that code, before toggling on the gate and allowing that data in.

pohly replied:
> If the gate is alpha and they opt-in and persist data which a 1.28 server considers invalid, that's on them. 

I agree - just wanted to be sure.

> sorry that wasn't clear

It was clear, just not whether downgrading to 1.28 must be supported.

pohly replied:
@liggitt: can you take another look? I think I have implemented the validation as intended.

pohly commented on pkg/apis/core/validation/validation.go:
```diff
@@ -5899,13 +5899,20 @@ func validateConfigMapNodeConfigSource(source *core.ConfigMapNodeConfigSource, f
+// Validate that the resource typename is qualified.
 // Refer to docs/design/resources.md for more details.
-func validateResourceName(value string, fldPath *field.Path) field.ErrorList {
+func validateResourceNameIsQualified(value string, fldPath *field.Path) field.ErrorList {
```
This code used to be part of `validateResourceName`. The call chains used to be:

- `validateContainerCommon` -> `ValidateResourceRequirements` -> `validateContainerResourceName` -> `validateResourceName`
- `ValidateResourceQuota` -> `ValidateResourceQuotaSpec` -> `ValidateResourceQuotaResourceName` -> `validateResourceName`

Therefore all names valid in a `ResourceQuota` (second chain) also had to be valid in a container (first chain). That's probably why "persistentvolumeclaims" is valid in a container although it makes no sense there.

By splitting up `validateResourceName`, `ValidateResourceQuotaResourceName` can call `validateResourceNameIsQualified` and then consult it's own, separate list via `IsStandardQuotaResourceName` - that check was already there before, it just got combined with `IsStandardResourceName` unnecessarily.

bart0sh commented:
/priority important-longterm

pohly commented:
Let's put this on hold for now (marked WIP), because in the light of ["semantic parameters"](https://github.com/kubernetes/enhancements/pull/4384#discussion_r1482068509) we might want to define a different quota mechanism.

pohly commented:
/reopen

I've been asked to reconsider doing quota by class via the v1.ResourceQuota...

pohly commented:
This includes https://github.com/kubernetes/kubernetes/pull/125488 for now. Only the last commit is really new and contains the changes of this PR.

We still need to decide whether we want this in 1.31.

johnbelamaric reviewed (commented):
Looks reasonable to me, though my confidence in reviewing the quota code and api is low.

The "all" behavior does beg for allocation-time enforcement, I can see that. But this is the best we can do at admission time and seems like a reasonable starting place.

thockin commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -6979,6 +6979,10 @@ const (
+	// resource.k8s.io ResourceClaims, number
+	ResourceClaims ResourceName = "resourceclaims"
+	// resource.k8s.io devices requested with a certain DeviceClass, number
+	ResourceClaimsPerClass string = ".deviceclass.resource.k8s.io/devices"
```
I don't love this but it has precedent in storageclass

thockin reviewed (commented):
API LGTM

thockin commented:
I only looked at API

/approve
/hold

deads2k reviewed (changes requested):
The idea is sound, but some updates are needed.

pohly commented on pkg/apis/core/validation/validation.go:
```diff
@@ -6331,14 +6338,25 @@ func validateContainerResourceName(value core.ResourceName, fldPath *field.Path)
-func ValidateResourceQuotaResourceName(value core.ResourceName, fldPath *field.Path) field.ErrorList {
-	allErrs := validateResourceName(value, fldPath)
+func ValidateResourceQuotaResourceName(value core.ResourceName, oldResources core.ResourceList, fldPath *field.Path) field.ErrorList {
+	allErrs := validateResourceNameIsQualified(value, fldPath)
```
`validateResourceName` used to check for "qualified" and "standard resource name" (= `helper.IsStandardResourceName`).  Then the code below in this function checked again with `IsStandardQuotaResourceName`.

I think that was redundant at best, but definitely confusing. Now this function here alone is responsible for checking which names without slash are allowed.

This is also where we can check the additional conditions for allowing the new quota names, depending on feature gates and whether they were set before.

pohly commented on pkg/apis/core/validation/validation.go:
```diff
@@ -6921,38 +6939,43 @@ func validateScopeSelector(resourceQuotaSpec *core.ResourceQuotaSpec, fld *field
 
-	allErrs = append(allErrs, ValidateResourceQuotaSpec(&resourceQuota.Spec, field.NewPath("spec"))...)
-	allErrs = append(allErrs, ValidateResourceQuotaStatus(&resourceQuota.Status, field.NewPath("status"))...)
+	allErrs = append(allErrs, ValidateResourceQuotaSpec(&resourceQuota.Spec, nil, field.NewPath("spec"))...)
+	allErrs = append(allErrs, ValidateResourceQuotaStatus(&resourceQuota.Status, nil, field.NewPath("status"))...)
```
So you are saying that `ValidateResourceQuotaSpec` should not get the additional `oldResourceQuotaSpec *core.ResourceQuotaSpec` argument? It then cannot be called as part of an update.

Instead, we need the same code again in a `ValidateResourceQuotaSpecUpdate` where the additional special case of "name was already before update" is considered.

Isn't it better to avoid that duplication of code and add the `oldResourceQuotaSpec *core.ResourceQuotaSpec` parameter, as done in this PR at the moment?

pohly replied:
BTW, I think `ValidateResourceQuotaStatus` is useless here. The status is always going to be empty when `ValidateResourceQuota` is being called for a new resource, right?

pohly replied:
> Isn't it better to avoid that duplication of code and add the oldResourceQuotaSpec *core.ResourceQuotaSpec parameter, as done in this PR at the moment?

Clarified on Slack: some code duplication is worth it because of the extra clarity.

pohly commented on pkg/quota/v1/evaluator/core/resource_claims.go:
```diff
@@ -0,0 +1,165 @@
+			// Worst case...
+			numDevices = resourceapi.AllocationResultsMaxSize
+		default:
+			// Shouldn't happen because other modes don't pass validation.
+			return nil, fmt.Errorf("%s: unsupported count mode", request.AllocationMode)
```
This code here runs in the apiserver. Therefore there is no version skew, this code must get update whenever a new AllocationMode gets added.

I can also add an "internal error" prefix to make this clearer.

pohly commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -6979,6 +6979,10 @@ const (
 	// Local ephemeral storage limit, in bytes. (500Gi = 500GiB = 500 * 1024 * 1024 * 1024)
 	ResourceLimitsEphemeralStorage ResourceName = "limits.ephemeral-storage"
+	// resource.k8s.io ResourceClaims, number
+	ResourceClaims ResourceName = "resourceclaims"
```
Okay (https://github.com/kubernetes/kubernetes/pull/120611/commits/e182d9227356156c56873881f7a3e6465fcdc631#r1688100489).

pohly replied:
And because of this removal, we don't need to change validation! Double win.

pohly commented on pkg/quota/v1/evaluator/core/resource_claims.go:
```diff
@@ -0,0 +1,165 @@
+	result[claimObjectCountName] = *(resource.NewQuantity(1, resource.DecimalSI))
+	for _, request := range claim.Spec.Devices.Requests {
+		if request.AdminAccess {
+			// No need to count, admins get unlimited access.
```
Probably that is what will happen. But I'm not sure whether it's always going to be that way, so it seemed better to add this special case.

pohly replied:
Permission to use "admin access" might not be tied to the namespace. It's checked via VAP, so there will be plenty of opportunities to configure it differently, potentially so that admins run pods with admin access to devices in a namespace where normal users are limited by quota.

pohly commented on pkg/quota/v1/evaluator/core/resource_claims.go:
```diff
@@ -0,0 +1,165 @@
+)
+
+// the name used for object count quota
+var claimObjectCountName = generic.ObjectCountQuotaResourceNameFor(resourceapi.SchemeGroupVersion.WithResource("resourceclaims").GroupResource())
```
Actually, this isn't the alias. It's the standard `count/resourceclaims.resource.k8s.io`.

When I tried to remove this, I found that because this evaluator registers itself for the `resourceclaims.resource.k8s.io` ResourceGroup, the normal code no longer handled that count. Therefore I am going to keep this.

pohly committed:
DRA e2e: fix the quota name

The actual name has the k8s.io suffix.

pohly commented on pkg/apis/core/validation/validation.go:
```diff
@@ -6292,13 +6292,20 @@ func validateConfigMapNodeConfigSource(source *core.ConfigMapNodeConfigSource, f
+// Validate that the resource typename is qualified.
 // Refer to docs/design/resources.md for more details.
-func validateResourceName(value core.ResourceName, fldPath *field.Path) field.ErrorList {
+func validateResourceNameIsQualified(value core.ResourceName, fldPath *field.Path) field.ErrorList {
```
So you are suggesting that instead of rejecting the feature gated names during validation, they should silently get removed the same way as gated fields are? And validation then doesn't need to check the feature gate?

I'm not sure whether that is better. When I discussed this with @liggitt [in 2023](https://github.com/kubernetes/kubernetes/pull/120611#discussion_r1368720017) I believe he wanted to check during validation.

pohly committed:
DRA quota: unit test case for resource.k8s.io quota names

The names aren't actually special for validation. They are
acceptable with and without the feature gate, the only difference
is that they don't do anything when the feature is enabled.

deads2k commented on pkg/quota/v1/evaluator/core/resource_claims.go:
```diff
@@ -0,0 +1,166 @@
+		default:
+			// Shouldn't happen because other modes don't pass validation.
+			// If a new mode gets added, this code needs to be updated.
+			return nil, fmt.Errorf("internal error: %s: unsupported count mode", request.AllocationMode)
```
This stills vulnerable to skew.

Oops scenario
1. v1.40 adds new DeviceAllocationMode="Awesome"
2. Forget to update this.
3. insta-fail

Downgrade scenario
1. v1.40 adds value
2. downgrade to 1.39
3. fail until removing resourceclaim

Standard upgrade skew
1. add in v1.40
2. start upgrade, 1.40 kube-apiserver (moves first), 1.39 kube-contorller-manager (uses evaluators to refill quota)
3. fail

pohly replied:
Ack. So your previous proposal to simply not count such requests makes sense. It's consistent with "new alpha fields don't work after a downgrade".

deads2k commented:
/lgtm
/approve

deads2k commented:
> I only looked at API
> 
> /approve /hold

/hold cancel

pacoxu commented:
New test cases that were added with this PR is failing in https://testgrid.k8s.io/sig-release-master-informing#capz-windows-master

pohly commented:
That test shouldn't run in that job because the test has the `[Feature:DynamicResourceAllocation]` label.

Looks like that job isn't configured correctly and should be fixed.

pohly commented:
'--ginkgo.focus=.... \[sig-api-machinery\].ResourceQuota|... ` that's probably it.

pacoxu commented:
> That test shouldn't run in that job because the test has the `[Feature:DynamicResourceAllocation]` label.
> 
> Looks like that job isn't configured correctly and should be fixed.

I opened https://github.com/kubernetes/kubernetes/issues/126321 to track the issue.
